End-to-end chain training

hhadian opened the pull request:
This PR is for review to decide whether to merge or not.
Currently, I have added phone-based and lexicon-free (character-based) recipes for WSJ only.

danpovey commented:
Thanks!  It may be a few days before I get to this.

vince62s commented:
That looks really interesting.
Do you think that in this simplified training recipe, the ngram order of the phone LM would have an impact ? btw did you keep 3 as the default ngram order ?

hhadian commented:
Thanks!
I am not sure about that because I didn't do any experiments regarding the denominator LM. But I plan to do a few basic experiments and I can share the results with you.
The default ngram order is 4 and the default no-prune ngram-order is 3 (similar to regular chain recipes).

galv commented on egs/wsj/s5/steps/libs/nnet3/train/chain_objf/acoustic_model.py:
```diff
@@ -355,7 +406,7 @@ def check_for_required_files(feat_dir, tree_dir, lat_dir):
              '{0}/final.mdl'.format(tree_dir), '{0}/tree'.format(tree_dir),
              '{0}/lat.1.gz'.format(lat_dir), '{0}/final.mdl'.format(lat_dir),
-             '{0}/num_jobs'.format(lat_dir), '{0}/splice_opts'.format(lat_dir)]
+             '{0}/num_jobs'.format(lat_dir)]
```
Why is splice_opts no longer required?

hhadian replied:
I guess it is not even required in regular chain training.

galv replied:
I prefer not to remove this, though, for the sake of backwards compatibility.

galv commented on src/gmmbin/Makefile:
```diff
@@ -5,10 +5,11 @@ EXTRA_CXXFLAGS = -Wno-sign-compare
 BINFILES = gmm-init-mono gmm-est gmm-acc-stats-ali gmm-align \
-           gmm-decode-faster gmm-decode-simple gmm-align-compiled \
-           gmm-sum-accs gmm-est-regtree-fmllr gmm-acc-stats-twofeats \
+           gmm-decode-faster gmm-decode-simple gmm-decode-nbest gmm-align-compiled \
```
Did you forget to add gmm-init-trans.cc and gmm-decode-nbest.cc?

hhadian replied:
I have made only one change in this file (i.e adding gmm-init-biphone)
The other changes must be due to a conflict.

galv commented on src/gmmbin/gmm-init-biphone.cc:
```diff
@@ -0,0 +1,245 @@
+
+    const std::vector<int32> &phones = topo.GetPhones();
+
+    std::vector<int32> phone2num_pdf_classes (1 + phones.back());
```
Remove whitespace here: `s (1`

galv replied:
What is this constructor doing? I don't understand it. Is the last element of phones guaranteed to be the largest phoneme id? Why are you adding plus 1? Is it because phones are indexed from 1 in Kaldi (not sure if that's true right this moment, but I recall something like that being true before).

hhadian replied:
Yes (to both questions).
BTW, such constructors are used in a couple of other places in Kaldi.

galv commented on src/gmmbin/gmm-init-biphone.cc:
```diff
@@ -0,0 +1,245 @@
+    ContextDependency *ctx_dep = NULL;
+    std::vector<std::vector<int32> > shared_phones;
+    if (shared_phones_rxfilename == "") {
+      for (size_t i = 0; i < phones.size(); i++)
```
Can you add braces to this for loop?

hhadian replied:
It's a single-line statement, why are suggesting "braces"?

galv replied:
Just a common best practice to always use curly braces, especially since your single-line statement (in my opinion) is rather long.

danpovey replied:
Actually the Google style guide doesn't mandate them for one-line statements, and in cases like this I prefer not to have them as they make the code longer.

galv commented on src/online2bin/extend-wav-with-silence.cc:
```diff
@@ -76,16 +76,47 @@ int main(int argc, char *argv[]) {
 
-    std::string wav_rspecifier = po.GetArg(1);
-    std::string wav_wspecifier = po.GetArg(2);
+    if (ClassifyRspecifier(po.GetArg(1), NULL, NULL) != kNoRspecifier) {
```
It's not very clear to me what you're doing here since it's so large. Since it looks like you made a new set of input options, can you change the usage string of this binary?

hhadian replied:
I am adding support for `rxfilename` and `wxfilename` so it can be used in a piped command.
I'll update the usage.

galv replied:
Great, thanks!

entn-at commented on egs/wsj/s5/local/run_end2end_char.sh:
```diff
@@ -0,0 +1,111 @@
+
+if [ $stage -le 5 ]; then
+  echo "$0: calling the flat-start chain recipe..."
+  loacl/chain/run_tdnn_lstm_flatstart.sh
```
`loacl` -> `local`

entn-at commented on egs/wsj/s5/local/run_end2end_phone.sh:
```diff
@@ -0,0 +1,93 @@
+
+if [ $stage -le 5 ]; then
+  echo "$0: calling the flat-start chain recipe..."
+  loacl/chain/run_tdnn_flatstart.sh
```
`loacl` -> `local`

hhadian commented on src/chain/chain-generic-numerator.cc:
```diff
@@ -0,0 +1,336 @@
+#include "chain/chain-kernels-ansi.h"
+
+namespace kaldi {
+namespace chain {
```
There is a short comment in the `.h` file but I will add more. 
This class (i.e. GenericNumeratorComputation) is responsible for Forward-Backward (to get the derivatives and total logprob) on a generic numerator graph (i.e. a training fst which can e.g. have self-loops).

hhadian commented on src/chain/chain-generic-numerator.cc:
```diff
@@ -0,0 +1,336 @@
+  const int32 num_sequences = supervision_.num_sequences,
+      num_states = max_num_hmm_states_;
+  // set alpha_0(0) for all sequences to 1.0 and leave the rest to be 0.0.
+  BFloat *first_frame_alpha = alpha_.RowData(0);
```
I defined `BFloat` (in the header file) so I could optionally change it to `double` to make some parts of computation double-precision. This was mostly meant for the initial experiments. I guess I can change it to BaseFloat now.

hhadian commented on src/chain/chain-generic-numerator.h:
```diff
@@ -0,0 +1,127 @@
+  // the derivs w.r.t. the nnet outputs (transposed)
+  Matrix<BaseFloat> nnet_output_deriv_transposed_;
+
+  // forward and backward probs matrices
```
Almost all the variables with the same name in this class and in `DenominatorComputation` (like  `alpha`, `beta`, `kMaxDerivTimeSteps`) have the exact same function. I will add some comments to make it clear.

hhadian commented on egs/wsj/s5/steps/libs/nnet3/train/chain_objf/acoustic_model.py:
```diff
@@ -49,9 +49,10 @@ def create_phone_lm(dir, tree_dir, run_opts, lm_opts=None):
-    common_lib.execute_command(
-        """copy-transition-model {tree_dir}/final.mdl \
-                {dir}/0.trans_mdl""".format(dir=dir, tree_dir=tree_dir))
+    if not os.path.exists('{dir}/0.trans_mdl'.format(dir=dir)):
```
I guess this change is for the initial experiments where I needed to copy my own transition model w/o being overwritten. I will probably need to revert this.

danpovey commented on egs/wsj/s5/steps/nnet3/chain/train_e2e.py:
```diff
@@ -0,0 +1,557 @@
+                        dest='left_deriv_truncate',
+                        default=None,
+                        help="Deprecated. Kept for back compatibility")
+    parser.add_argument("--chain.apply-deriv-weights", type=str,
```
a couple of args here seem to be repeated.

danpovey commented:
@vimalmanohar, do you have time to give this a quick check before I merge?

vimalmanohar commented:
Sure. I will go through it.

On Fri, Jan 19, 2018 at 6:42 PM Daniel Povey <notifications@github.com>
wrote:

> @vimalmanohar <https://github.com/vimalmanohar>, do you have time to give
> this a quick check before I merge?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/2072#issuecomment-359120309>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEATV3Q91CH8O_sx2V3POsO8Qv3DjK5Pks5tMShrgaJpZM4Q7rgT>
> .
>


-- 
Vimal Manohar
PhD Student
Electrical & Computer Engineering
Johns Hopkins University

vimalmanohar commented on egs/wsj/s5/local/chain/e2e/run_tdnn_flatstart.sh:
```diff
@@ -0,0 +1,223 @@
+
+# training options
+num_epochs=4.5
+initial_effective_lrate=0.001
```
I think a lot of options are presented here. Some of them can be fixed and removed if they are not too important for tuning.

vimalmanohar commented on egs/wsj/s5/local/chain/e2e/run_tdnn_flatstart.sh:
```diff
@@ -0,0 +1,223 @@
+      for lmtype in tgpr bd_tgpr; do
+        steps/nnet3/decode.sh \
+          --acwt 1.0 --post-decode-acwt 10.0 \
+          --extra-left-context $chunk_left_context \
```
left and right contexts can be removed for tdnn scripts.

vimalmanohar commented on egs/wsj/s5/local/chain/e2e/run_tdnn_flatstart.sh:
```diff
@@ -0,0 +1,223 @@
+shared_phones=true
+train_set=train_si284_spe2e_hires
+test_sets="test_dev93 test_eval92"
+first_layer_splice=-1,0,1
```
I think such things can be fixed.

hhadian replied:
I agree.

vimalmanohar commented on egs/wsj/s5/local/e2e/run_end2end_char.sh:
```diff
@@ -0,0 +1,111 @@
+if [ $stage -le 4 ]; then
+  echo "$0: estimating character language model for the denominator graph"
+  mkdir -p exp/chain/e2e_base
+  cat data/$trainset/text | \
```
This should probably be run using $train_cmd as systems might not be configured to run such things locally.

vimalmanohar commented on egs/wsj/s5/local/e2e/run_end2end_phone.sh:
```diff
@@ -0,0 +1,93 @@
+#!/bin/sh
+# Copyright 2017    Hossein Hadian
+
```
Mention what this script is and also maybe mention similar scripts like _char.sh etc.

vimalmanohar commented on egs/wsj/s5/local/e2e/run_end2end_phone.sh:
```diff
@@ -0,0 +1,93 @@
+
+if [ $stage -le 2 ]; then
+  echo "$0: perturbing the training data to allowed lengths"
+  utils/data/get_utt2dur.sh data/$trainset  # necessary for next command
```
Perhaps you can wrap all these steps into a script like utils/data/perturb_data_dir_lengths.sh or some thing like that.

hhadian replied:
I've tried to minimize introducing new "shared" scripts-- at least in the first PR.
@danpovey, what would you suggest?

vimalmanohar commented on egs/wsj/s5/steps/libs/nnet3/train/chain_objf/acoustic_model.py:
```diff
@@ -120,6 +120,56 @@ def generate_chain_egs(dir, data, lat_dir, egs_dir,
+                           left_context_initial=-1, right_context_final=-1,
+                           frame_subsampling_factor=3,
+                           alignment_subsampling_factor=3,
+                           feat_type='raw', online_ivector_dir=None,
```
Only feat_type raw is now supported. So don't give this as an option.

vimalmanohar commented on egs/wsj/s5/utils/data/perturb_speed_to_allowed_lengths.py:
```diff
@@ -0,0 +1,246 @@
+import copy
+import math
+
+parser = argparse.ArgumentParser(description="""This script copies the 'srcdir'
```
This script can be modified to follow the normal conventions like in train.py using main() and get_args(). Google standards does not allow global variables in general etc.

hhadian replied:
will fix this (and other scripts).

vimalmanohar commented on egs/wsj/s5/utils/data/perturb_speed_to_allowed_lengths.py:
```diff
@@ -0,0 +1,246 @@
+                                        math.log(factor))))
+
+
+# 1b. compute and write allowed lengths
```
All this should be inside a function. Also add doc strings.

vimalmanohar commented on egs/wsj/s5/utils/data/segment_data.sh:
```diff
@@ -0,0 +1,33 @@
+. ./path.sh
+
+if [ $# != 2 ]; then
+  echo "Usage: "
```
This needs a clear usage message. 
Also, the name of this file is confusing since it is not segmenting data. @danpovey might have a better name.

hhadian replied:
Actually, I've been meaning to ask for name suggestions. Maybe `extract_segments_data_dir.sh`?
The other script is text_to_phones.py (which converts word-level transcriptions to phone-level transcriptions). Maybe `word_transcription_to_phone_transcription.py` ?

vimalmanohar commented on egs/wsj/s5/utils/data/segment_data.sh:
```diff
@@ -0,0 +1,33 @@
+set -e
+set -o pipefail
+
+extract-segments scp:$srcdir/wav.scp $srcdir/segments ark,scp:$dir/data/segments.ark,$dir/segments.scp
```
I think this can be parallelized since it can get slow for large data. Also change the name from segments.ark to wav_segments.ark for clarity.

vimalmanohar commented on egs/wsj/s5/utils/data/segment_data.sh:
```diff
@@ -0,0 +1,33 @@
+set -o pipefail
+
+extract-segments scp:$srcdir/wav.scp $srcdir/segments ark,scp:$dir/data/segments.ark,$dir/segments.scp
+cat $dir/segments.scp | awk '{ print $1 " wav-copy " $2 " - |" }' >$dir/wav.scp
```
I don't think you need wav-copy here. Programs can directly read from ark files.

vimalmanohar commented on egs/wsj/s5/utils/text_to_phones.py:
```diff
@@ -0,0 +1,84 @@
+import math
+import random
+
+parser = argparse.ArgumentParser(description="""This script reads
```
This can also be put into functions, but since this is a small script it may not be essential.

vimalmanohar commented on src/chain/chain-generic-numerator.h:
```diff
@@ -0,0 +1,173 @@
+#include "matrix/kaldi-matrix.h"
+#include "hmm/transition-model.h"
+#include "chain/chain-supervision.h"
+#include "cudamatrix/cu-matrix.h"
```
I think you are including files that are not required or can be included in cc files.

hhadian replied:
Yes, I guess these were for some unusual experiments.

vimalmanohar commented on src/chain/chain-training.cc:
```diff
@@ -53,21 +57,50 @@ void ComputeChainObjfAndDeriv(const ChainTrainingOptions &opts,
+  } else {
+    GenericNumeratorComputation numerator(supervision, nnet_output);
+    num_logprob_weighted = numerator.Forward();
+    KALDI_LOG << "Numerator logprob per frame: "
```
I guess this was for debugging. You can probably remove it

vimalmanohar commented on src/chain/chain-training.cc:
```diff
@@ -53,21 +57,50 @@ void ComputeChainObjfAndDeriv(const ChainTrainingOptions &opts,
+    KALDI_LOG << "Numerator logprob per frame: "
+              << num_logprob_weighted / (*weight);
+    num_ok = (num_logprob_weighted - num_logprob_weighted == 0);
+    KALDI_LOG << "Numerator Forward " << (num_ok ? "succeeded" : "failed");
```
Also remove this. You can probably keep track of how many failed and print somewhere else. But logging all this would be very slow. At least use like KALDI_VLOG(2) or something.

hhadian replied:
I wouldn't worry about speed [it happens only once for each minibatch], but it might be unnecessary.
I could use VLOG.

danpovey replied:
I agree that there should be a count maintained.  If you expect it to normally succeed, maybe you could print something out if it failed?
Also I don't like the name `num_ok`, it's ambiguous.    I prefer `numerator_ok`.

hhadian replied:
It never fails for WSJ. It could fail a few times (for long utterances, only during the first 10 or so iterations) on harder databases (like SWBD). I guess printing something if it fails is enough if you agree? 
I guess maintaining a count involves changing a few different functions/classes (like  ObjectiveFunctionInfo, UpdateStats, and even ComputeChainObjfAndDeriv). Should I do it?

I will fix the name `num_ok`.

danpovey replied:
No, don't change those classes, since they interact with other code.
But there's something else that I'd like you to change, that I just noticed.
Your changes are making the function ComputeChainObjfAndDeriv quite complicated.
I'd like you to, in chain-training.cc but not the header, introduce a new function
ComputeChainObjfAndDerivE2e, and if supervision.e2e is true, call that function.
That will keep the main current code path simple and free of your changes.

hhadian replied:
Will do.

misbullah commented:
Hi,

I address some bugs in the following files:

utils/data/perturb_speed_to_allowed_lengths.py

line 212: 
original:     u1.wavefile = '{} sox -t wav - -t wav - speed {} | '.format(u.wavefile, speed)
need to change to be:     u1.wavefile = 'sox -t wav {} -t wav - speed {} | '.format(u.wavefile, speed)

line 230: 
original:  u2.wavefile = '{} sox -t wav -t wav - speed {} | '.format(u.wavefile, speed)
need to change to be:  u2.wavefile = 'sox -t wav {} -t wav - speed {} | '.format(u.wavefile, speed)

line 241L
original:     u3.wavefile = '{} extend-wav-with-silence --extra-silence-length={} - - | '.format(u.wavefile, delta)
need to change to be: u3.wavefile = 'extend-wav-with-silence --extra-silence-length={} {} - | '.format(delta, u.wavefile)

In steps/libs/nnet3/train/chain_objf/acoustic_model.py

line 138: 
original: steps/nnet3/chain/get_egs_e2e.sh
need to change to be: steps/nnet3/chain/e2e/get_egs_e2e.sh

I also found that there is no --add-deltas option for --egs.opts "--normalize-egs true --add-deltas false" in local/chain/e2e/run_tdnn_flatstart.sh

Thanks
Alim

hhadian commented:
Thanks, I've already fixed all of them (and other issues) and I'll push them soon.

hhadian commented:
I addressed the issues. I guess this is ready.
@danpovey, all the scripts directly dealing with e2e are in separate directories but there are 3 other shared scripts (`text_to_phones.py`, `perturb_speed_to_allowed_lengths.py`, and `extract_wav_segments_data_dir.sh`) which are in `utils` and `utils/data`. Should I move them to e.g. `utils/e2e` too?

danpovey commented:
Thanks, will try to look at this today or to-morrow.


On Sat, Jan 27, 2018 at 3:53 PM, Hossein Hadian <notifications@github.com>
wrote:

> I addressed the issues. I guess this is ready.
> @danpovey <https://github.com/danpovey>, all the scripts directly dealing
> with e2e are in separate directories but there are 3 other shared scripts (
> text_to_phones.py, perturb_speed_to_allowed_lengths.py, and
> extract_wav_segments_data_dir.sh) which are in utils and utils/data.
> Should I move them to e.g. utils/e2e too?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/2072#issuecomment-361014473>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu7xBs4A-f60nSd8D0ts6uKTgfiBgks5tO4zLgaJpZM4Q7rgT>
> .
>

danpovey commented on src/chain/chain-supervision.cc:
```diff
@@ -603,7 +683,8 @@ int32 ComputeFstStateTimes(const fst::StdVectorFst &fst,
+    e2e(other.e2e), e2e_fsts(other.e2e_fsts) { }
 
 void AppendSupervision(const std::vector<const Supervision*> &input,
                        bool compactify,
```
These functions AppendSupervision() and AddWeightToSupervisionFst() have become very long due to this e2e stuff.  We try to avoid having functions longer than about 40 lines.  I think the best way to keep the two parts of the code separate would be to have a statement near the beginning of those functions that checks if the 'e2e' flag is set, and if it is, recurses to a static function with a name like `AppendSupervisionE2e()`.

danpovey commented on src/chain/chain-supervision.cc:
```diff
@@ -616,70 +697,111 @@ void AppendSupervision(const std::vector<const Supervision*> &input,
-  KALDI_ASSERT(supervision->fst.Properties(fst::kIEpsilons, true) == 0);
-  return true;
+  if (!supervision->e2e) {
+    // remove epsilons before composing.  'normalization_fst' has noepsilons so
```
noepsilons -> no epsilons. remove -> Remove.  Watch capitalization of comments.

danpovey commented on src/chain/chain-supervision.h:
```diff
@@ -256,6 +269,19 @@ bool ProtoSupervisionToSupervision(
+    it would return false.
+    To find out more about end-to-end training, see chain-generic-numerator.h
+ */
+bool TrainingGraphToSupervision(
```
I think you should add 'E2e' to the end of this function name to indicate that it relates to the e2e setup.

danpovey commented on src/chain/chain-supervision.h:
```diff
@@ -256,6 +269,19 @@ bool ProtoSupervisionToSupervision(
+/** This function creates and initializes an end-to-end supervision object
+    from a training FST (e.g. created using compile-train-graphs). It simply
+    sets all the input and output labels to pdf_id+1 (i.e. converts the FST to
+    an FSA) and stores the resulting FST in e2e-fsts[0].
```
I think by `e2e-fsts[0]` you mean mean `supervision->e2e_fsts[0]`.

danpovey commented on src/chain/chain-supervision.h:
```diff
@@ -256,6 +269,19 @@ bool ProtoSupervisionToSupervision(
+    from a training FST (e.g. created using compile-train-graphs). It simply
+    sets all the input and output labels to pdf_id+1 (i.e. converts the FST to
+    an FSA) and stores the resulting FST in e2e-fsts[0].
+    It returns true if there were no epsilon transitions, otherwise
```
maybe explain why it's a problem if there were epsilons.

danpovey commented on src/gmmbin/gmm-init-biphone.cc:
```diff
@@ -0,0 +1,245 @@
+        level1_map[pset[k]] = new TableEventMap(0, level2_map); //lvl2_eventmap;
+        if (k != pset.size() - 1)
+          current_pdfid = base_pdfid;
+      } /////////   k
```
strange comment

danpovey commented on src/gmmbin/gmm-init-biphone.cc:
```diff
@@ -0,0 +1,245 @@
+      po.PrintUsage();
+      exit(1);
+    }
+
```
this file seems to have a bunch of whitespace lines that are probably excessive.

danpovey commented on src/gmmbin/gmm-init-biphone.cc:
```diff
@@ -0,0 +1,245 @@
+
+    const std::vector<int32> &phones = topo.GetPhones();
+
+    std::vector<int32> phone2num_pdf_classes (1 + phones.back());
```
remove space before (.

danpovey commented on egs/wsj/s5/utils/data/perturb_speed_to_allowed_lengths.py:
```diff
@@ -0,0 +1,298 @@
+            u3 = copy.deepcopy(u)
+            u3.id = 'pv3-' + u.id
+            u3.speaker = 'pv3-' + u.speaker
+            u3.wavefile = '{} extend-wav-with-silence --extra-silence-length={} - - | '.format(u.wavefile, delta)
```
is there a reason you can't just use 'sox pad'?

danpovey commented on egs/wsj/s5/utils/text_to_phones.py:
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python
```
this script is kind of specialized so I'd rather not put it here-- please move it to one of the 'e2e' directories in steps/.

danpovey commented on src/chain/chain-supervision.cc:
```diff
@@ -550,18 +610,38 @@ void Supervision::Read(std::istream &is, bool binary) {
+        delete compact_fst;
+      }
+    }
+    ExpectToken(is, binary, "</FSTs>");
```
I'd prefer these tokens to be called Fsts rather than FSTs.

danpovey commented on src/gmmbin/gmm-init-biphone.cc:
```diff
@@ -0,0 +1,245 @@
+
+
+    std::string topo_filename = po.GetArg(1);
+    int dim = atoi(po.GetArg(2).c_str());
```
And please use ConvertStringToInteger, e.g. `if (!ConvertStringToInteger(...) || dim <= 0 || dim > 10000) KALDI_ERR...`

vimalmanohar commented on egs/wsj/s5/utils/data/perturb_speed_to_allowed_lengths.py:
```diff
@@ -0,0 +1,298 @@
+        if utt in text and utt in utt2dur and utt in utt2spk:
+            utterances.append(Utterance(utt, wav_scp[utt], utt2spk[utt],
+                                  text[utt], utt2dur[utt]))
+        else:
```
Try to add some count like num_success, num_fail and print them. Don't print warning for every utterance. There should also be a special check that the input data directory has been passed through extract_wav_segments script. Otherwise it would generate a lot of warnings without any information on what actually is the issue. 
In additional also check how many utterances you got compared to the ones in wav_scp, and exit if its zero or less than some percentage.

vimalmanohar commented on egs/wsj/s5/utils/data/extract_wav_segments_data_dir.sh:
```diff
@@ -0,0 +1,37 @@
+
+if [ $# != 2 ]; then
+  echo "Usage: $0 <srcdir> <destdir>"
+  echo " This script copies data directory <srcdir> to <destdir> and gets"
```
This script will be very slow. Split the data and run the command in parallel.

danpovey replied:
If you do ad --nj and --cmd options to this script, don't set the default of --nj to more than 4, for I/O reasons.

vimalmanohar commented on egs/wsj/s5/utils/data/extract_wav_segments_data_dir.sh:
```diff
@@ -0,0 +1,37 @@
+set -e -o pipefail
+utils/copy_data_dir.sh $srcdir $dir
+
+extract-segments scp:$srcdir/wav.scp $srcdir/segments ark,scp:$dir/data/wav_segments.ark,$dir/segments.scp
```
Better to not put the temporary files like $dir/segments.scp in $dir. Keep it in $dir/data.
Better to rename it to wav_segments.scp to avoid confusion with segments file.

vimalmanohar commented on egs/wsj/s5/utils/data/perturb_speed_to_allowed_lengths.py:
```diff
@@ -0,0 +1,298 @@
+                        help="""Chain frame subsampling factor.
+                             See steps/nnet3/chain/train.py""")
+    parser.add_argument('--speed-perturb',  type=bool, default=True,
+                        help="""If false, no speed perturbation will occur, i.e.
```
Type casting to bool will change "false" to True. You can instead type=str, choices=['true','false'] and then convert the string to bool yourself.

vimalmanohar commented:
Yes nj=4 and cmd=run.pl can be the default.

On Fri, Feb 9, 2018 at 1:44 PM Daniel Povey <notifications@github.com>
wrote:

> *@danpovey* commented on this pull request.
> ------------------------------
>
> In egs/wsj/s5/utils/data/extract_wav_segments_data_dir.sh
> <https://github.com/kaldi-asr/kaldi/pull/2072#discussion_r167312436>:
>
> > @@ -0,0 +1,37 @@
> +#!/bin/bash
> +
> +# Copyright    2017  Hossein Hadian
> +# Apache 2.0
> +
> +# This script copies a data directory (which has a 'segments' file), extracting
> +# wav segments (according to the 'segments' file)
> +# so that the resulting data directory does not have a 'segments' file anymore.
> +
> +. utils/parse_options.sh
> +. ./path.sh
> +
> +if [ $# != 2 ]; then
> +  echo "Usage: $0 <srcdir> <destdir>"
> +  echo " This script copies data directory <srcdir> to <destdir> and gets"
>
> If you do ad --nj and --cmd options to this script, don't set the default
> of --nj to more than 4, for I/O reasons.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/2072#discussion_r167312436>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEATV6mXxXAf8LLBVT9cJCDwR6bry535ks5tTJIfgaJpZM4Q7rgT>
> .
>
-- 
Vimal Manohar
PhD Student
Electrical & Computer Engineering
Johns Hopkins University

vimalmanohar commented on egs/wsj/s5/utils/data/perturb_speed_to_allowed_lengths.py:
```diff
@@ -0,0 +1,298 @@
+            ## Add two versions for the second allowed_duration
+            ## one version is by using speed modification using sox
+            ## the other is by extending by silence
+            if not args.speed_perturb:
```
There should be no `not` here.

danpovey commented:
Hm, I thought this had been merged already.  It seems not.  I think there may be un-handled comments still though, and I see a conflict now.

hhadian commented:
This is almost ready (have not pushed yet). Two notes:
I didn't know there was a `sox pad`. I checked it now, but it seems that - unlike `extend-wav-with-silence` - it does not use the silence at the begin/end to make the padding (i.e. it just inserts some samples with 0 amplitude).

About `extract_wav_segments.sh`, I guess a better way than make it use parallel jobs is to change it so that it does not do any actual wave processing and instead writes `sox trim` commands in wav.scp with the segment start/end times taken from the 'segments' file. It will be faster and it will save disk space. What would you think?

danpovey commented:
Regarding extract_wav_segments.sh:

Using "sox trim" will be extremely wasteful of disk I/O when the time comes
to actually use the files, because it will have to read the entire file
from disk for each segment.


On Sat, Mar 3, 2018 at 1:05 PM, Hossein Hadian <notifications@github.com>
wrote:

> This is almost ready (have not pushed yet). Two notes:
> I didn't know there was a sox pad. I checked it now, but it seems that -
> unlike extend-wav-with-silence - it does not use the silence at the
> begin/end to make the padding (i.e. it just inserts some samples with 0
> amplitude).
>
> About extract_wav_segments.sh, I guess a better way than make it use
> parallel jobs is to change it so that it does not do any actual wave
> processing and instead writes sox trim commands in wav.scp with the
> segment start/end times taken from the 'segments' file. It will be faster
> and it will save disk space. What would you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/2072#issuecomment-370167490>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu1Rh8jfvs_J4XNIBN2PX7egdKoPSks5tatttgaJpZM4Q7rgT>
> .
>

hhadian commented:
All issues are addressed.
